File: tests/agent_workloads/adapters/mcp_resource.py
import json
import os
import subprocess
import sys


if len(sys.argv) != 2:
    raise SystemExit("usage: mcp_resource INPUT_ROOT")

root = os.path.abspath(sys.argv[1])
task = os.environ["JET_CORPUS_TASK"]
project = os.path.join(root, "denied" if task == "mcp-environment-denied" else "readonly")
root_uri = "file://" + project


def frame(body):
    return f"Content-Length: {len(body.encode())}\r\n\r\n{body}"


bodies = [
    json.dumps({"jsonrpc": "2.0", "id": 1, "method": "initialize", "params": {"rootUri": root_uri, "capabilities": {}}}, separators=(",", ":")),
    '{"jsonrpc":"2.0","method":"notifications/initialized","params":{}}',
    '{"jsonrpc":"2.0","id":2,"method":"resources/list","params":{}}',
    '{"jsonrpc":"2.0","id":3,"method":"resources/read","params":{"uri":"jet://environment"}}',
    '{"jsonrpc":"2.0","id":4,"method":"resources/read","params":{"uri":"jet://missing"}}',
    '{"jsonrpc":"2.0","id":5,"method":"shutdown","params":{}}',
    '{"jsonrpc":"2.0","method":"exit","params":{}}',
]
payload = "".join(frame(body) for body in bodies)
completed = subprocess.run(
    [os.environ["JET_CORPUS_JET"], "self", "lsp"],
    cwd=project,
    env={**os.environ, "HOME": "/agent-secret"},
    input=payload,
    text=True,
    capture_output=True,
    timeout=10,
    check=False,
)
if completed.returncode != 0:
    raise SystemExit(completed.stderr or "MCP server failed")
response = completed.stdout
if task == "mcp-environment-denied":
    if "-32002" not in response:
        raise SystemExit("MCP denied case did not fail closed")
    if "/agent-secret" in response:
        raise SystemExit("MCP denied case leaked HOME")
    print("mcp=denied\nerror=-32002\nsecret=redacted")
else:
    for marker in ("active_environment", "codex-agent", "mcp.read", "generated.txt", "lint"):
        if marker not in response:
            raise SystemExit(f"MCP resource omitted {marker}")
    if "/agent-secret" in response:
        raise SystemExit("MCP resource leaked HOME")
    print("mcp=readonly\nresource=jet://environment\ngrant=mcp.read\nside-effects=none")

File: tests/agent_workloads/adapters/mcp_resource.mjs
import { spawnSync } from "node:child_process";
import { resolve } from "node:path";

if (process.argv.length !== 3) {
  throw new Error("usage: mcp_resource INPUT_ROOT");
}

const root = resolve(process.argv[2]);
const task = process.env.JET_CORPUS_TASK;
const project = `${root}/${task === "mcp-environment-denied" ? "denied" : "readonly"}`;
const rootUri = `file://${project}`;
function frame(body) {
  return `Content-Length: ${Buffer.byteLength(body)}\r\n\r\n${body}`;
}
const payload = [
  JSON.stringify({ jsonrpc: "2.0", id: 1, method: "initialize", params: { rootUri, capabilities: {} } }),
  '{"jsonrpc":"2.0","method":"notifications/initialized","params":{}}',
  '{"jsonrpc":"2.0","id":2,"method":"resources/list","params":{}}',
  '{"jsonrpc":"2.0","id":3,"method":"resources/read","params":{"uri":"jet://environment"}}',
  '{"jsonrpc":"2.0","id":4,"method":"resources/read","params":{"uri":"jet://missing"}}',
  '{"jsonrpc":"2.0","id":5,"method":"shutdown","params":{}}',
  '{"jsonrpc":"2.0","method":"exit","params":{}}',
].map(frame).join("");
const result = spawnSync(process.env.JET_CORPUS_JET, ["self", "lsp"], {
  cwd: project,
  env: { ...process.env, HOME: "/agent-secret" },
  input: payload,
  encoding: "utf8",
  timeout: 10000,
});
if (result.error) throw result.error;
if (result.status !== 0) throw new Error(result.stderr || "MCP server failed");
const response = result.stdout;
if (task === "mcp-environment-denied") {
  if (!response.includes("-32002")) throw new Error("MCP denied case did not fail closed");
  if (response.includes("/agent-secret")) throw new Error("MCP denied case leaked HOME");
  console.log("mcp=denied\nerror=-32002\nsecret=redacted");
} else {
  for (const marker of ["active_environment", "codex-agent", "mcp.read", "generated.txt", "lint"]) {
    if (!response.includes(marker)) throw new Error(`MCP resource omitted ${marker}`);
  }
  if (response.includes("/agent-secret")) throw new Error("MCP resource leaked HOME");
  console.log("mcp=readonly\nresource=jet://environment\ngrant=mcp.read\nside-effects=none");
}
